### ISSUE TITLE
Performance improvement suggestion

### ISSUE THREAD

**Issue by struggggle** [samuelclay/NewsBlur#977]
Dear developers,
I found several bitmap displaying issues in NewsBlur. These issues could affect performance. 

(1) In the position:
com.newsblur.util.PrefsUtils.java  getUserImage()
the bitmap object is created with the invocation of decodeFile().
I found getUserImage() is invoked directly and indirectly by:
com.newsblur.fragment.ProfileDetailsFragment.java   onCreateView()
com.newsblur.activity.Main.java                     onCreate()
com.newsblur.fragment.LoginProgressFragment$LoginTask.java    onPostExecute()
which means image decoding operation is performed in the UI thread.
Android documentation says this:
https://developer.android.com/training/displaying-bitmaps/process-bitmap.html
" The BitmapFactory.decode* methods should not be executed on the main UI thread if the source data is read from disk or a network location"
Image decoding is slow. So, do you think the slow operation will affect NewsBlur's performance (e.g., make GUI not smooth enough). If yes, for performance considerations, we should perform the image decoding operations in worker threads (e.g., via AsyncTask).

In addition, Android documentation says this:
https://developer.android.com/training/displaying-bitmaps/load-bitmap.html#load-bitmap
" Images come in all shapes and sizes. In many cases they are larger than required for a typical application user interface (UI)"
" To avoid java.lang.OutOfMemory exceptions, check the dimensions of a bitmap before decoding it, unless you absolutely trust the source to provide you with predictably sized image data that comfortably fits within the available memory "
Do you think we should check the dimensions of a bitmap before decoding it? If the bitmap is larger than required, we can resize it before decoding, so as to reduce memory and time overhead.

(2) In the position:
com.newsblur.util.ImageLoader.java  decodeBitmap()
the bitmap object is created with the invocation of decodeFile().
I found decodeBitmap() is invoked indirectly by: 
com.newsblur.view.ActivityDetailsAdapter.java    getView()
which means that there will be many bitmap objects being created. 
Do you think we should also check the dimensions of a bitmap before decoding it in this position?

In addition, we can use Option.inbitmap to reuse the memory space of existing bitmap objects, which will reduce memory consumption and improves app's performance.

Looking forward to your response and hope I can help improve NewsBlur.
Thanks.

References: 
The use of Option.inbitmap:
https://www.youtube.com/watch?v=_ioFW3cyRV0&list=PLWz5rJ2EKKc9CBxr3BVjPTPoDPLdPIFCE&index=54
https://developer.android.com/training/displaying-bitmaps/manage-memory.html

**Comment by dosiecki** [comment]
Do you have any call timing data or allocation logs to support the notion that these calls are problematic?  We have somewhat extensively profiled the app and neither of these have ever showed up as taking nontrivial time.  Alternatively, are you experiencing some particular lag or excess resource usage you would like us to look into?  If so, please copy and paste the debug info you can get by using the "Send app feedback" link in the menu of the app.

**Comment by struggggle** [comment]
Hello, @dosiecki 
I don't have call timing data or allocation logs. The improvement suggestions come from the detection results of my static code analysis tool, which is designed to detect whether an app exists some space for performance improvement. 
In practical use, when I use NewsBlur to browse "GLOBAL SHARED STORIES" and fast slid down, I found some delays in images displaying (a few images need to wait several seconds to display). I checked NewsBlur's source code and found these positions that we can do some optimizations to make NewsBlur better.

**Comment by dosiecki** [comment]
In this case, the tool has flagged the wrong thing. All loading of the dynamic images when browsing through stories is done properly off of the main thread.  The only things that happen during lifecycle steps are extraordinarily small, static assets.

If you are getting extreme lag while scrolling through any story list, definitely copy and paste in the debug info from the feedback page as mentioned above.  I'll need to know which options you have enabled and what sort of hardware you are running on to dig into this much deeper.

**Comment by struggggle** [comment]
@dosiecki 
Yes, All loading of the dynamic images when browsing through stories is done off of the main thread. I know you've done this optimization. 

I would like to use an example to illustrate my point:

When browsing "GLOBAL SHARED STORIES", I can see the page below:
![image](https://cloud.githubusercontent.com/assets/19150666/21351308/da4f0288-c6f6-11e6-8336-fb65f41d8100.png)
 
When the item "Zinke" was opened, I can see this page:
![image](https://cloud.githubusercontent.com/assets/19150666/21351317/e20f572a-c6f6-11e6-8f78-92f0a94433ec.png)
 
I found both pages showed the same image. The link to the image is: http://www.lawyersgunsmoneyblog.com/wp-content/uploads/2016/12/ZinkeSportsmen.jpeg,
and the original size of the image is 515KB (1200*675)

Obviously, the size of the image is larger than required (Because we only need to display a thumbnail that reaches a certain clarity). If we directly load and display the image, which will require extra time and memory overhead and affect app's performance. In addition, there will be many images to be loaded and displayed when I browsing other stories.

So I suggest checking the dimensions of an image before loading it. If the image is larger than required, we can load a scaled down version, so as to reduce time and memory overhead.

Also, we can use BitmapFactory.Options.inBitmap to reuse the memory space of existing bitmap objects, which will greatly reduce the memory consumption and improves app's performance.

**Comment by dosiecki** [comment]
> Obviously, the size of the image is larger than required (Because we only need to display a thumbnail that reaches a certain clarity). If we directly load and display the image, which will require extra time and memory overhead and affect app's performance. In addition, there will be many images to be loaded and displayed when I browsing other stories.

This is behaving as designed.  From the outset, we have no way of knowing what images a website will publish, or how large they will be; we can only try our best to display what sites publish.  Also, image thumbnails are intended to be a rich feature for users with modern devices.  If you have an older device that struggles to load stories smoothly when thumbnails are on, I would definitely recommend turning off the feature via the main preferences menu.

> So I suggest checking the dimensions of an image before loading it. If the image is larger than required, we can load a scaled down version, so as to reduce time and memory overhead.

I don't have the time budget for this project to get too deep into the specifics, but this is an approach that we tried and ditched.  Any value achieved by trying to HEAD an image from a 3rd party is negated by the unreliably poor way that content-lengths are reported.  Not to mention, even if we know the image is huge, we still have to download the whole thing to display it.  We do cache the image after first download, but trying to also cache the resized version adds additional disk IO to the process, and on many devices, cache IO is more constrained than memory.  We try to cache as aggressively as possible while still being a polite consumer of device storage an memory. (you can find the exact formulas for how much of each we might use by perusing the code)

> Also, we can use BitmapFactory.Options.inBitmap to reuse the memory space of existing bitmap objects, which will greatly reduce the memory consumption and improves app's performance.

The catch here is that we would have to hold onto an enormous reference collection that maps image URLs onto the bitmap objects.  We generate the bitmaps as efficiently as possible and hand them to the layout engine for use, but our goal is to generate as little overhead as possible outside of the layout engine so that it has as much headroom as possible.  Coming up with a way to reference track the intermediate objects we use in between download and handoff to the layout engine would be expensive in both resources and technical overhead.

In summary, yes, image preview thumbnails are extremely expensive to enable.  They were designed as a rich media feature for devices with plenty of resources.  We could absolutely keep optimizing the way they are loaded, little bit little, if @samuelclay wants to budget the development time and code complexity.

**Comment by samuelclay** [comment]
Yeah I don't want to budget that in, since the iOS solution was to load the full size image and just display it. On my test devices I don't see any issues related to thumbnail loading.

**Comment by struggggle** [comment]
@dosiecki @samuelclay Thanks for your detailed reply.
Actually, we can know the size of an image before we decode it to be a bitmap.
You can see the details here:
https://developer.android.com/training/displaying-bitmaps/load-bitmap.html
"Setting the inJustDecodeBounds property to true while decoding avoids memory allocation, returning null for the bitmap object but setting outWidth, outHeight and outMimeType. This technique allows you to read the dimensions and type of the image data prior to construction (and memory allocation) of the bitmap."
